storage: return one entry less in Entries

tbg opened the pull request:
This works around the bug outlined in:

https://github.com/etcd-io/etcd/pull/10063

by matching Raft's internal implementation of commit pagination.
Once the above PR lands, we can revert this commit (but I assume
that it will take a little bit), and I think we should do that
because the code hasn't gotten any nicer to look at.

Fixes #28918.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/29579)
<!-- Reviewable:end -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though this is worth having another pair of eyes scrutinize it. @nvanbenschoten?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29579#-:-LLePrsq9QeDuEsa2Bfo:brqnnhb)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/entry_cache.go, line 196 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29579#-LLePCMi3-aOl6Ps39MR:-LLePCMj2cFOYN2ZXnJL:byc9hl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a051ed9f0c03f48aeb671859d6d584e2f683aa84/pkg/storage/entry_cache.go#L196)):*
> ```Go
> 		ent := v.(*raftpb.Entry)
> 		size := uint64(ent.Size())
> 		if maxBytes > 0 && bytes+size > maxBytes {
> ```

Did you consider an alternative approach to add a `truncateEntries()` function which takes a slice of entries, `lo` and `hi` indexes and a `maxSize` setting and duplicates the etcd/raft semantics? That seems like it would avoid a change here and could be more easily tested.


<!-- Sent from Reviewable.io -->

nvb reviewed (dismissed):
> Reviewed 5 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29579)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/storage/entry_cache.go, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29579#-LLeVox7-Ax2HUhNy7Wl:-LLeVox7-Ax2HUhNy7Wm:b-gs4slr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a051ed9f0c03f48aeb671859d6d584e2f683aa84/pkg/storage/entry_cache.go#L181)):*
> ```Go
> func (rec *raftEntryCache) getEntries(
> 	ents []raftpb.Entry, rangeID roachpb.RangeID, lo, hi, maxBytes uint64,
> ) (_ []raftpb.Entry, size uint64, nextIndex uint64, hitMaxSize bool) {
> ```

`s/hitMaxSize/exceededMaxBytes/`
___
*[pkg/storage/entry_cache.go, line 196 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29579#-LLePCMi3-aOl6Ps39MR:-LLeWBz63ovrZA_bYELd:bxa4z7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/a051ed9f0c03f48aeb671859d6d584e2f683aa84/pkg/storage/entry_cache.go#L196)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Did you consider an alternative approach to add a `truncateEntries()` function which takes a slice of entries, `lo` and `hi` indexes and a `maxSize` setting and duplicates the etcd/raft semantics? That seems like it would avoid a change here and could be more easily tested.
</blockquote></details>

If we did that, we'd need to make sure we call it twice so that we can short-circuit before hitting RocksDB in the happy-path. If we're doing that then it doesn't seem to buy us much.
___
*[pkg/storage/replica_raftstorage.go, line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29579#-LLeWy_O25zh-0Alx_tn:-LLeWy_P14bW0Jq4dNaA:b-gixfuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a051ed9f0c03f48aeb671859d6d584e2f683aa84/pkg/storage/replica_raftstorage.go#L162)):*
> ```Go
> 		// Note that we track the size of proposals with payloads inlined.
> 		size += uint64(ent.Size())
> 		if maxBytes > 0 && size > maxBytes {
> ```

What does it mean for `maxBytes == 0`? I thought Raft uses `math.MaxUint64` when it doesn't want a limit.


<!-- Sent from Reviewable.io -->

tbg committed:
storage: return one entry less in Entries

This works around the bug outlined in:

https://github.com/etcd-io/etcd/pull/10063

by matching Raft's internal implementation of commit pagination.
Once the above PR lands, we can revert this commit (but I assume
that it will take a little bit), and I think we should do that
because the code hasn't gotten any nicer to look at.

Fixes #28918.

Release note: None

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29579)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/entry_cache.go, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29579#-LLeVox7-Ax2HUhNy7Wl:-LLiXCZG4H3nxtkchOI9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a051ed9f0c03f48aeb671859d6d584e2f683aa84/pkg/storage/entry_cache.go#L181)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/hitMaxSize/exceededMaxBytes/`
</blockquote></details>

Done.
___
*[pkg/storage/entry_cache.go, line 196 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29579#-LLePCMi3-aOl6Ps39MR:-LLiXIJc7_aAcTXh7qc8:b-cuusrj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a051ed9f0c03f48aeb671859d6d584e2f683aa84/pkg/storage/entry_cache.go#L196)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If we did that, we'd need to make sure we call it twice so that we can short-circuit before hitting RocksDB in the happy-path. If we're doing that then it doesn't seem to buy us much.
</blockquote></details>

I did consider it, but as Nathan pointed out it didn't look any better.
___
*[pkg/storage/replica_raftstorage.go, line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29579#-LLeWy_O25zh-0Alx_tn:-LLinZQZExS_7Dc0N8xY:b-to21rq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a051ed9f0c03f48aeb671859d6d584e2f683aa84/pkg/storage/replica_raftstorage.go#L162)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What does it mean for `maxBytes == 0`? I thought Raft uses `math.MaxUint64` when it doesn't want a limit.
</blockquote></details>

I think this was just crufty. Removed.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29579)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/entry_cache.go, line 196 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29579#-LLePCMi3-aOl6Ps39MR:-LLj50RZ6jwQxF4ITPGX:b-rrqhs6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a051ed9f0c03f48aeb671859d6d584e2f683aa84/pkg/storage/entry_cache.go#L196)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I did consider it, but as Nathan pointed out it didn't look any better.
</blockquote></details>

Right, we'd have to call it twice, but we'd ensure that the same logic is used in both places and could test it in isolation. I don't feel strongly about this and you're already testing both logic paths, so carry on.


<!-- Sent from Reviewable.io -->

tschottdorf commented:
bors r=petermattis

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29579#-:-LLjK7467hqRpzz5VCj6:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)



<!-- Sent from Reviewable.io -->
